src/V3Const.cpp

```diff
@@ -96,18 +96,33 @@ class ConstBitOpTreeVisitor final : public AstNVisitor {
         ConstBitOpTreeVisitor* m_parentp;  // ConstBitOpTreeVisitor that holds this VarInfo
         AstVarRef* m_refp;  // Points the variable that this VarInfo covers
         V3Number m_bitPolarity;  // Coefficient of each bit
-        static int widthOfRef(AstVarRef* refp) {
-            if (AstWordSel* selp = VN_CAST(refp->backp(), WordSel)) return selp->width();
-            if (AstCCast* castp = VN_CAST(refp->backp(), CCast)) return castp->width();
-            return refp->width();
-        }
 
     public:
         // METHODS
         bool hasConstantResult() const { return m_constResult >= 0; }
         bool sameVarAs(const AstNodeVarRef* otherp) const { return m_refp->sameGateTree(otherp); }
         void setPolarity(bool compBit, int bit) {
             UASSERT_OBJ(!hasConstantResult(), m_refp, "Already has result of " << m_constResult);
+            UASSERT_OBJ(bit < VL_QUADSIZE, m_refp,
+                        "bit:" << bit << " is too big after V3Expand"
+                               << " back:" << m_refp->backp());
+            if (bit >= m_bitPolarity.width()) {
+                const V3Number oldPol = std::move(m_bitPolarity);
+                // oldPol.width() is 8, 16, or 32 because this visitor is called after V3Expand
+                const int newWidth = oldPol.width() * 2;
```

Comment:
What is the * 2 for?  Below you are only initing up to width, not width*2.

Reply:
The code is expanding m_bitPolarity.
So `newWidth` must be greater than `bit + 1`.
Expanding to `bit + 1` is maybe ok, but expanding by 2x is better because
- CCast will cast to such bitwidth anyway (33 bit width has to be 64 in C++)
- can avoid frequent expansion

The expanded `m_bitPolarity` is filled by 'X', then copy from `oldPol` .
https://github.com/verilator/verilator/blob/6af31c3305f630757d931fbfcdaac8df04bf6f53/src/V3Const.cpp#L120-L126

This is matches the initialization inf ctor.
https://github.com/verilator/verilator/blob/6af31c3305f630757d931fbfcdaac8df04bf6f53/src/V3Const.cpp#L179-L183


Thanks for highlighting this code. I noticed that `m_bitPolarity` has to be expanded more than 2x in some cases. (e.g. 8 to 32)
So I pushed 6af31c3305f630757d931fbfcdaac8df04bf6f53.

Reply:
@wsnyder Does the comment above answer your question ?

Reply:
Yes, thanks